### ISSUE TITLE
Too many requests

### ISSUE THREAD

**Issue by Anwesh1** [PathwayCommons/pubmed-dl#3]
I'm basically getting all the PMID's then trying to pass it in as a loop to pubmed-lookup.
The specific code for it is down here: (store has all the PMID's) 

```python
  for i in store:
        url = 'http://www.ncbi.nlm.nih.gov/pubmed/'+i
        lookup = PubMedLookup(url, email)
        publication = Publication(lookup)
        ab.write(publication.title)
```
this is a picture of the error as well.
![Screenshot 2021-01-19 160203](https://user-images.githubusercontent.com/57552053/105092788-b507f780-5a6f-11eb-9f3b-091242ee83fe.png)

**Comment by JohnGiorgi** [comment]
The access limit for Eutils (with a key) is 10 PMIDs per second (see @jvwong 's comment below). This loop is (likely) making >10 requests per second so this error is expected.

You will have to limit the # of requests per second artificially. Can you try the following

```python
import time

# Set this to 3 if you are using an API key and 10 otherwise
doc_per_sec = 10

for i, pmid in enumerate(store):
    url = f"http://www.ncbi.nlm.nih.gov/pubmed/{pmid}"
    lookup = PubMedLookup(url, email)
    publication = Publication(lookup)
    ab.write(publication.title)
    if (i + 1) % doc_per_sec == 0:
        time.sleep(1)  # You should only have to wait one second before the next batch of ten requests
```

I doubt this is the most elegant solution but let's see if it gets us past this error. Then we can compare this to your existing solution.

**Comment by jvwong** [comment]
https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/

> Now that I have a key, are there still access limits?

> Yes. By default, your key will increase the limit to 10 requests/second for all activity from that key. If we receive requests at a higher frequency that include the same key, all requests using that key will receive an error message. If you need higher rates of access, please contact us and we can discuss your situation.

**Comment by jvwong** [comment]
@Anwesh1 did you take a look at this: https://github.com/PathwayCommons/semantic-search/blob/master/semantic_search/ncbi.py

**Comment by Anwesh1** [comment]
> @Anwesh1 did you take a look at this: https://github.com/PathwayCommons/semantic-search/blob/master/semantic_search/ncbi.py

I did actually, but I was unable to pinpoint what that script is specifically doing (I also tried to run it but that didn't help).

**Comment by jvwong** [comment]
The 'public' method in that module is:

```python
# -- Public methods --
def uids_to_docs(uids: List[str]) -> List[Dict[str, Collection[Any]]]:
    """Return uid, and text (i.e. title + abstract) given a PubMed uid"""
    ...
```
The idea is it takes a list of PubMed `uids` like this:

```json
["9887103", "30049242", "22936248"]
```

Grabs all the article data in batches of 10 000, and returns a list of dicts:

```python
{ 
  "uid": <PubMed Id>,
  "text": <title + abstract>
}
```

It has the capability of taking in an access token, updating the ESEARCH parameters etc..

**Comment by Anwesh1** [comment]
On the pubmed-lookup, I don't get the 'too many requests' error once I mod with 3. However, there is a weird error being thrown after working on a few records. For example, I have 10 PMID's, and I'm grabbing the title for each of these. I am able to get it for about 4-5 of them then I'm getting an error with the library that says this (screenshot): 
![image](https://user-images.githubusercontent.com/57552053/105234255-fadcc280-5b38-11eb-847c-9ef6cf8abf8d.png)

@jvwong I see, I will try to work with it by passing it some PMID's and try to grab its data, thanks

**Comment by Anwesh1** [comment]
@jvwong Question on the ncbi.py script. You said that I need to pass a list to the uids_to_docs() 'public' method right? Or is there something else I need to pass with it? because I'm getting an error that says
`Error encountered in uids_to_docs Invalid URL '': No schema supplied. Perhaps you meant http://?`
Please clarify if possible. thanks

**Comment by jvwong** [comment]
Can you give me a bit information about how you tried to use it so I can try to reproduce what you did and the error?

**Comment by Anwesh1** [comment]
I just made a basic small list and tried to call the function with that list, it was just 2 lines.
```python
def main():
    store = ["33453708", "33453707", "33453706", "33453705"]
    uids_to_docs(store)
```

**Comment by jvwong** [comment]
My best guess is that the module is trying to find environment variable values set in the `Settings` class and declared in the  accompanying `.env` file in that repository. ` e.g. `EUTILS_BASE_URL`.

**Comment by Anwesh1** [comment]
all the environment variable values set in `settings` class are empty, I'm pretty sure. Is there something I do to change anything?

**Comment by JohnGiorgi** [comment]
> On the pubmed-lookup, I don't get the 'too many requests' error once I mod with 3. However, there is a weird error being thrown after working on a few records. For example, I have 10 PMID's, and I'm grabbing the title for each of these. I am able to get it for about 4-5 of them then I'm getting an error with the library that says this (screenshot):
> ![image](https://user-images.githubusercontent.com/57552053/105234255-fadcc280-5b38-11eb-847c-9ef6cf8abf8d.png)
> 
> @jvwong I see, I will try to work with it by passing it some PMID's and try to grab its data, thanks

I would have to see the exact code you ran to help debug thus. At some point `get` (a `dict` method) is being called on a `str`.

**Comment by JohnGiorgi** [comment]
@Anwesh1 @jvwong On a broader note, if @jvwong has already written all the logic to get article data from pmids with eutils, maybe we should just pull it out of semantic search and place it in this repository. 

Then we can maintain that code here separately from semantic search and use it for other projects.

**Comment by Anwesh1** [comment]
@JohnGiorgi @jvwong I see, I guess that is something we can discuss further tomorrow on the call. Another thing is that I just ran my script for all the data produced in December of 2020 (124081 records). 

I pulled all the data according to the format we talked about (PMID\tTITLE\tABSTRACT). All of these records took 494 seconds to run completely. I'm pretty sure if we use `pubmed-lookup()` it would run considerably slowly as we would need to place a 1 second buffer every 10 records.
